tests/unit/threaded_deque_test.c

```diff
@@ -0,0 +1,426 @@
+#include <test.h>
+
+#include <alloc.h>
+#include <mutex.h>
+#include <threaded_deque.h>
+
+/* Memory illustration legend:          *
+ *      | : memory bounds               *
+ *      > : left                        *
+ *      < : right                       *
+ *      ^ : left + right (empty)        *
+ *      v : left + right (at capacity)  *
+ *      x : used memory                 *
+ *      - : unused memory               */
+
+static void test_push_pop(void)
+{
+    // Initialised with DEFAULT_CAPACITY = 16
+    ThreadedDeque *deque = ThreadedDequeNew(0, free);
+    // |^---------------|
+    ThreadedDequePushLeft(deque, xstrdup("1"));
+    // |<-------------->|
```

Comment:
@karlhto I still find this thoroughly confusing. After 1 push, shouldn't `left == right`, since a PopLeft or a PopRight would both pop the same index?

Reply:
Alright, so:

The right index is the first empty element in a continuous block of data (that can wrap around the allocated memory area). I made it this way so I wouldn't have to have a special case when `size` is `0`, where after 

I also think doing it differently will mess with the internals of the array. It makes sense to me that you should end up at different indices when popping from the left and from the right, e.g.:
```c
/* Starting position (with one element in the data structure), where:
 *     < is right index (first empty element)
 *     - is empty
 *     > is left index (first non-empty element)
 *     ^ is both indexes when array is empty
 */
|<-------------->|

// Pop left:
|^---------------|

// Pop right
|---------------^|
```

Since we already have the `size` parameter, it is not a big deal to change.

Reply:
I think the legend is missing the information that `<` is actually a first **empty** element while `>` is a **non-empty** element.